src/network/connection.js

```diff
@@ -292,28 +302,50 @@ module.exports = class Connection {
           this.failIfNotConnected()
           const entry = { apiKey, apiName, apiVersion, correlationId, resolve, reject }
 
+          const sendRequest = () => {
```

Comment:
I'm not sure that this is the place to handle this. I think it should be inside the requestQueue instead. Unless I'm reading this wrong, the request queue will now think that a request has been sent even if it's actually being throttled. Conceptually, it feels like it makes more sense for the requestQueue to be the one to do the actual throttling by simply pushing the request onto the queue instead of executing it directly, and then not executing any more requests until the throttle time has passed - same as if `maxInFlightRequests` has been exceeded. From reading the Java implementation, that looks more or less how they've implemented it (`NetworkClient.maybeThrottle`).

Since the requestQueue doesn't decode the response, it still makes sense for the connection to be the one to actually decide for how long to throttle and based on what, but the actual enforcement of the throttling should be moved into the requestQueue.

Reply:
I can try that. 

When looking at this yesterday the request queue approach looked like considerable more intrusive changes, while putting into the actual connection was basically trivial and nicely symmetric.

Reply:
It is, but putting it there has a serious drawback, which is that if you are throttled, any requests you make will look like they are "completed" even though they have actually just been scheduled. If you make subsequent requests for which you don't expect a response, the "send" call will resolve immediately even though the request has not been made. Potentially you could even pile up requests in this limbo state. Using the request queue makes it very clear that the request has not been made yet, and the "send" call will not resolve until the throttling is over and the request has actually been made.

Reply:
Ok, the changes were indeed a bit more intrusive, but ultimately I agree that putting this in the request queue feels more "right" now.